subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ivy/IvyDynamicRevisionRemoteResolveIntegrationTest.groovy

```diff
@@ -108,6 +108,263 @@ dependencies {
+                compile group: "group", name: "projectB", version: "latest.release"
+            }
+
+          import javax.inject.Inject
+     
+          class MP implements ComponentMetadataSupplier {
+          
+            final RepositoryResourceAccessor repositoryResourceAccessor
+            
+            @Inject
+            MP(RepositoryResourceAccessor accessor) { repositoryResourceAccessor = accessor }
+            
+            int calls
+            Map<String, String> status = [:]
+          
+            void supply(ComponentMetadataSupplierDetails details) {
+                def id = details.id
+                println "Providing metadata for \$id"
+                repositoryResourceAccessor.withResource("status.txt") {
+                    if (status.isEmpty()) {
+                        println "Parsing status file call count: \${++calls}"
+                        it.withReader { reader ->
+                            reader.eachLine { line ->
+                                if (line) {
+                                   def (module, st) = line.split(';')
+                                   status[module] = st
+                                }
+                            }
+                        }
+                        println status
+                    }
+                }
+                details.result.status = status[id.toString()]
+            }
+          }
+"""
+        when:
+        def projectA1 = ivyHttpRepo.module("group", "projectA", "1.1").publish()
+        def projectA2 = ivyHttpRepo.module("group", "projectA", "1.2").publish()
+        def projectB1 = ivyHttpRepo.module("group", "projectB", "1.1").publish()
+        def projectB2 = ivyHttpRepo.module("group", "projectB", "2.2").publish()
+        ivyHttpRepo.module("group", "projectA", "2.0").publish()
+
+
+        and:
+        def statusFile = temporaryFolder.createFile("versions.status")
+        statusFile << '''group:projectA:1.1;release
+group:projectA:1.2;release
+group:projectB:1.1;release
+group:projectB:2.2;integration
+'''
+        server.expectGet("/repo/status.txt", statusFile)
+        expectGetDynamicRevision(projectA2)
+        expectGetDynamicRevision(projectB1)
+
+        then: "custom metadata rule prevented parsing of ivy descriptor"
+        checkResolve "group:projectA:1.+": "group:projectA:1.2", "group:projectB:latest.release": "group:projectB:1.1"
+        outputContains 'Providing metadata for group:projectB:2.2'
+        outputContains 'Providing metadata for group:projectB:1.1'
+        !output.contains('Providing metadata for group:projectA:1.1')
+
+        and: "remote status file parsed only once"
+        outputContains 'Parsing status file call count: 1'
+        !output.contains('Parsing status file call count: 2')
+
+        when: "resolving the same dependencies"
+        server.expectHead("/repo/status.txt", statusFile)
+        server.expectHead("/repo/status.txt", statusFile)
```

Comment:
I understand why this is happening. When we call `MetadataProvider#getComponentMetadata`, we avoid calling `resolve()` if possible, because of the metadata supplier. If there's one, it's called first, and we return the metadata, which is **not** a `BuildableModuleComponentMetaDataResolveResult` (because we don't need all metadata, we need a small portion of it).

As a consequence, when a supplier answers, we return directly, and perform no caching (in-memory, or persistent). This means that the supplier is always going to be called. The only thing which is cached is the remote call, and that's why we're seeing `HEAD` requests.